File: backend/internal/chat/adapters/httpapi/sources.go
package httpapi

import (
	"net/http"

	"github.com/corsi/backend/internal/chat/app"
	"github.com/corsi/backend/internal/chat/domain"
	"github.com/corsi/backend/internal/platform/render"
)

// Agent Sources over HTTP.
//
// Same nesting as memories, and for the same reason: a source belongs to
// its agent, so there is no address for one that does not name it.
//
// One asymmetry worth stating: the list does NOT carry `content`. A source
// is up to twenty thousand characters and a list is a list — it returns
// titles, sizes and state, and `GET .../sources/{sourceId}` returns the
// document when something actually needs to read it.

type createSourceRequest struct {
	Title       string `json:"title"`
	Description string `json:"description"`
	Content     string `json:"content"`
}

func (h *Handler) createSource(w http.ResponseWriter, r *http.Request) {
	ws, ok := h.workspaceID(w, r)
	if !ok {
		return
	}
	agentID, ok := h.pathID(w, r)
	if !ok {
		return
	}
	var req createSourceRequest
	if !h.decode(w, r, &req) {
		return
	}
	src, err := h.svc.CreateSource(r.Context(), app.CreateSourceInput{
		WorkspaceID: ws,
		AgentID:     agentID,
		Title:       req.Title,
		Description: req.Description,
		Content:     req.Content,
	})
	if err != nil {
		h.writeDomainErr(w, err)
		return
	}
	render.JSON(w, http.StatusCreated, src)
}

type updateSourceRequest struct {
	Title       *string `json:"title"`
	Description *string `json:"description"`
	Content     *string `json:"content"`
	Enabled     *bool   `json:"enabled"`
}

func (h *Handler) updateSource(w http.ResponseWriter, r *http.Request) {
	ws, ok := h.workspaceID(w, r)
	if !ok {
		return
	}
	agentID, ok := h.pathID(w, r)
	if !ok {
		return
	}
	id, ok := h.pathUUID(w, r, "sourceId")
	if !ok {
		return
	}
	var req updateSourceRequest
	if !h.decode(w, r, &req) {
		return
	}
	src, err := h.svc.UpdateSource(r.Context(), app.UpdateSourceInput{
		WorkspaceID: ws,
		AgentID:     agentID,
		ID:          id,
		Title:       req.Title,
		Description: req.Description,
		Content:     req.Content,
		Enabled:     req.Enabled,
	})
	if err != nil {
		h.writeDomainErr(w, err)
		return
	}
	render.JSON(w, http.StatusOK, src)
}

func (h *Handler) getSource(w http.ResponseWriter, r *http.Request) {
	ws, ok := h.workspaceID(w, r)
	if !ok {
		return
	}
	agentID, ok := h.pathID(w, r)
	if !ok {
		return
	}
	id, ok := h.pathUUID(w, r, "sourceId")
	if !ok {
		return
	}
	src, err := h.svc.GetSource(r.Context(), ws, agentID, id)
	if err != nil {
		h.writeDomainErr(w, err)
		return
	}
	render.JSON(w, http.StatusOK, src)
}

func (h *Handler) deleteSource(w http.ResponseWriter, r *http.Request) {
	ws, ok := h.workspaceID(w, r)
	if !ok {
		return
	}
	agentID, ok := h.pathID(w, r)
	if !ok {
		return
	}
	id, ok := h.pathUUID(w, r, "sourceId")
	if !ok {
		return
	}
	if err := h.svc.DeleteSource(r.Context(), ws, agentID, id); err != nil {
		h.writeDomainErr(w, err)
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// sourceListEnvelope mirrors the memory one: the list shape plus the three
// numbers the page needs to explain itself. `total` above `limit` means the
// read was capped, which is the only way a cap is allowed to happen.
type sourceListEnvelope struct {
	Items            []domain.Source `json:"items"`
	Limit            int             `json:"limit"`
	Total            int64           `json:"total"`
	UsedCharacters   int             `json:"used_characters"`
	BudgetCharacters int             `json:"budget_characters"`
}

func (h *Handler) listSources(w http.ResponseWriter, r *http.Request) {
	ws, ok := h.workspaceID(w, r)
	if !ok {
		return
	}
	agentID, ok := h.pathID(w, r)
	if !ok {
		return
	}
	page, err := h.svc.ListSources(r.Context(), ws, agentID)
	if err != nil {
		h.writeDomainErr(w, err)
		return
	}
	render.JSON(w, http.StatusOK, sourceListEnvelope{
		Items:            page.Items,
		Limit:            page.Limit,
		Total:            page.Total,
		UsedCharacters:   page.UsedCharacters,
		BudgetCharacters: page.BudgetCharacters,
	})
}

File: frontend/src/modules/agents/api/sources.ts
/**
 * Typed REST client for /chat/agents/{id}/sources.
 *
 * A Source is reference material the agent can consult: a named piece of
 * text the user wrote or pasted on purpose. It is not Memory — see
 * `@/modules/agents/api/memories` for the other side of that boundary.
 *
 * ── The list does not carry the text ───────────────────────────────────
 * A source runs to twenty thousand characters, so `listSources` returns
 * titles, sizes and state with `content` empty, and `getSource` returns the
 * document when the editor opens it. Content can never legitimately be
 * empty (the database rejects it), so an empty string here unambiguously
 * means "not fetched".
 */

import { apiFetch } from "@/lib/api/client";

export interface ApiSource {
  id: string;
  workspace_id: string;
  agent_id: string;
  title: string;
  /** A note for the human. Never sent to the model. */
  description: string;
  /** Empty in list responses. See the note above. */
  content: string;
  enabled: boolean;
  /** Exact rune count of the content, measured server-side. */
  characters: number;
  /** `characters` through the module's one heuristic. An estimate. */
  estimated_tokens: number;
  /**
   * Whether this source would reach the model on the next turn. Computed by
   * the backend with the same function that composes the turn.
   */
  in_context: boolean;
  /**
   * This source cannot fit the block on its own, whatever else is switched
   * off. A different fact from `in_context: false`, and shown differently:
   * one is "not this time", the other is "not ever, until you shorten it".
   */
  oversized: boolean;
  created_at: string;
  updated_at: string;
}

export interface SourcePage {
  items: ApiSource[];
  /** The cap the server applied. `total` above it means the list was cut. */
  limit: number;
  total: number;
  /** What the selected sources contribute to a turn, header included. */
  used_characters: number;
  budget_characters: number;
}

export interface CreateSourceRequest {
  title: string;
  description?: string;
  content: string;
}

export type UpdateSourceRequest = Partial<CreateSourceRequest> & { enabled?: boolean };

export function listSources(agentId: string): Promise<SourcePage> {
  return apiFetch<SourcePage>(`/chat/agents/${agentId}/sources`);
}

export function getSource(agentId: string, id: string): Promise<ApiSource> {
  return apiFetch<ApiSource>(`/chat/agents/${agentId}/sources/${id}`);
}

export function createSource(agentId: string, body: CreateSourceRequest): Promise<ApiSource> {
  return apiFetch<ApiSource>(`/chat/agents/${agentId}/sources`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function updateSource(
  agentId: string,
  id: string,
  body: UpdateSourceRequest,
): Promise<ApiSource> {
  return apiFetch<ApiSource>(`/chat/agents/${agentId}/sources/${id}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

export function deleteSource(agentId: string, id: string): Promise<void> {
  return apiFetch<void>(`/chat/agents/${agentId}/sources/${id}`, { method: "DELETE" });
}

/** Mirrors the backend CHECK constraints, so the counters agree. */
export const MAX_SOURCE_TITLE = 120;
export const MAX_SOURCE_DESCRIPTION = 280;
export const MAX_SOURCE_CONTENT = 20000;

/**
 * The module's token heuristic, for text that has not been saved yet.
 *
 * This is the one place the rule is duplicated outside Go, and it is
 * duplicated for exactly one reason: the editor shows the size of a draft
 * while it is being typed, and a draft does not exist on the server to be
 * measured. **Every saved source reports `estimated_tokens` from the
 * backend** — never recompute it here for a record that came back from the
 * API.
 *
 * It must stay identical to `app.EstimateTokens`: ceil(characters / 4),
 * measured in characters and never in bytes. It is an estimate, it is named
 * one, and it must be presented as one.
 */
export function estimateTokens(characters: number): number {
  if (characters <= 0) return 0;
  return Math.ceil(characters / 4);
}

/** Rune count, matching how the backend and the database both measure. */
export function countCharacters(text: string): number {
  return [...text].length;
}
